### PR TITLE
Extract to validator

### PR DESCRIPTION
**Pull request by Matt-Warnock**

### CODE REVIEW

File: bin/app [hunk 1]
```diff
@@ -8,9 +8,11 @@ require 'controler'
 require 'creator'
 require 'null_action'
 require 'user_interface'
+require 'validator'
 
 array_database = ArrayDatabase.new
-user_interface = UserInterface.new($stdin, $stdout)
+validator = Validator.new
```
Comment (octopusinvitro) [inline bin/app:14]: No need to create a variable for this, you can instantiate it directly in the argument of the UI

File: lib/user_interface.rb [hunk 2]
```diff
@@ -33,27 +33,26 @@ class UserInterface
     notes: 'Contact notes: '
   }.freeze
 
-  VALID_YES_NO_REPLY = /^[yn]$/i.freeze
```
Comment (octopusinvitro) [inline lib/user_interface.rb:36]: The text is knowledge of the UI, I would let this constant here.

File: lib/validator.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+class Validator
+  VALID_YES_NO_REPLY = /^[yn]$/i.freeze
+
+  def valid_choice?(option)
+    option.match?(/^[1-2]$/)
```
Comment (octopusinvitro) [inline lib/validator.rb:7]: I would leave this knowledge with the UI as you had it before.
It's not the responsibility of the validator to know how many choices the menu has.

File: lib/validator.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  end
+
+  def valid_field?(field, value)
+    {
+      phone: valid_phone?(value),
+      email: valid_email?(value)
+    }.fetch(field, true)
+  end
+
+  def valid_phone?(value)
```
Comment (octopusinvitro) [inline lib/validator.rb:17]: `valid_phone?` and `valid_email?` should be private, there is no need to expose them.
The UI only needs to know about `valid_field?`, and nobody else is using these two methods.
Expose as little as you can always.
You can test these two methods through the `valid_field?` method.

File: lib/validator.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    value.match?(/^\d{11}$/)
+  end
+
+  def valid_email?(value)
+    value.match?(/@/)
+  end
+
+  def valid_yes_no_answer?(value)
+    value.match?(VALID_YES_NO_REPLY)
```
Comment (octopusinvitro) [inline lib/validator.rb:26]: get this constant from the UI

File: spec/controler_spec.rb [hunk 6]
```diff
@@ -2,15 +2,17 @@
 
 require 'controler'
 require 'null_action'
+require 'validator'
 
 RSpec.describe Controler do
   let(:actions) { [NullAction.new, NullAction.new] }
   let(:exit_choice) { UserInterface::EXIT_CHOICE }
   let(:output) { StringIO.new }
+  let(:validator) { Validator.new }
```
Comment (octopusinvitro) [inline spec/controler_spec.rb:11]: you can just instantiate when needed, no need to create another variable in this case

File: spec/validator_spec.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+require 'validator'
+
+RSpec.describe Validator do
+  let(:validator) { described_class.new }
+  describe '#valid_choice?' do
```
Comment (octopusinvitro) [inline spec/validator_spec.rb:7]: leave an empty line between these two

File: spec/validator_spec.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    let(:highest_choice_input) { 2 }
+    it 'returns true on avaible choices' do
```
Comment (octopusinvitro) [inline spec/validator_spec.rb:9]: empty line

File: spec/validator_spec.rb [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+    it 'returns false on non-avaible choices' do
+      non_choice_input = (highest_choice_input + 1).to_s
+
+      result = validator.valid_choice?(non_choice_input)
+
+      expect(result).to eq(false)
+    end
+  end
+  describe '#valid_field?' do
```
Comment (octopusinvitro) [inline spec/validator_spec.rb:23]: empty line

File: spec/validator_spec.rb [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+      expect(result).to eq(true)
+    end
+
+    it 'returns true by defalt on non-specified validation methods' do
+      result = validator.valid_field?(:notes, 'I think he has an Oscar')
+
+      expect(result).to eq(true)
+    end
+  end
+  describe '#valid_phone?' do
```
Comment (octopusinvitro) [inline spec/validator_spec.rb:36]: Move all the tests that you have in the valid phone describe and all the tests that you have in the valid email describe inside of the valid fields describe.

Then delete the first test you have in the valid fields describe.

File: spec/validator_spec.rb [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    end
+
+    it 'returns false with an invalid email address' do
+      expect(validator.valid_email?('matt.yahoo.com')).to eq(false)
+    end
+  end
+
+  describe '#valid_yes_no_answer?' do
+    it 'returns true on a valid yes or no reponse by user' do
+      expect(validator.valid_yes_no_answer?('y')).to eq(true)
```
Comment (octopusinvitro) [inline spec/validator_spec.rb:58]: Use the UI constant for `'y'`

You can add the case sensitivity test after this test.

File: spec/user_interface_spec.rb [hunk 12]
```diff
@@ -36,7 +38,7 @@
 
     it 'validates input, vaild numbers only' do
```
Comment (octopusinvitro) [inline spec/user_interface_spec.rb:39]: You can remove this test as it is already tested in the test in line 57

File: spec/user_interface_spec.rb [hunk 13]
```diff
@@ -54,23 +56,23 @@
...
 
     it 'returns a valid input' do
-      ui = described_class.new(valid_input, output)
+      ui = described_class.new(valid_input, output, validator)
 
       expect(ui.menu_choice).to eq(exit_choice)
     end
   end
 
   describe '#ask ask_for_fields' do
```
Comment (octopusinvitro) [inline spec/user_interface_spec.rb:73]: typo `ask ask_for_fields`

ask is repeated twice

File: spec/user_interface_spec.rb [hunk 14]
```diff
@@ -160,7 +162,7 @@
 
     it 'ignores case sensitivity on valid inputs' do
```
Comment (octopusinvitro) [inline spec/user_interface_spec.rb:163]: I would move this test to the validator, as case sensitivity is now knowledge that belongs to the validator.

**Review by octopusinvitro — CHANGES_REQUESTED** [review]
Just a comment on the extraction:

you were cutting the methods from the UI and pasting them into the validator without injecting the validator in the UI until the very end.
 
This means that your UI was literally broken during the whole process of extraction. The UI tests were red.

You usually either create the class first and then inject it in the original class and only then delete the methods, or inject it from the beginning and move the methods one by one as you did.

File: spec/user_interface_spec.rb (hunk 14 above)
Comment (Matt-Warnock) [reply spec/user_interface_spec.rb:163]: ? 
the constant is in lowercase

**Review by octopusinvitro — APPROVED** [review]

**Pull request merged**